src/herder/SurgePricingUtils.cpp

```diff
@@ -41,10 +50,315 @@ feeRate3WayCompare(int64 lFeeBid, uint32 lNbOps, int64 rFeeBid, uint32 rNbOps)
+            {
+                erase(currIt);
+                continue;
+            }
+            else
+            {
+                break;
+            }
+        }
+
+        // At this point, `currIt` points at the top transaction in the queue
+        // that is still within operation limits, so we can visit it and remove
+        // it from the queue.
+        auto const& txStack = *currIt;
+        auto visitRes = visitor(*txStack);
+        auto ops = txStack->getTopTx()->getNumOperations();
+        // Only account for operation counts when transaction was actually
+        // processed by the visitor.
+        if (visitRes == VisitTxStackResult::TX_PROCESSED)
+        {
+            opsLeftUntilLimit -= ops;
+        }
+        // Pop the tx from current iterator, unless the whole tx stack is
+        // skipped.
+        if (visitRes == VisitTxStackResult::TX_STACK_SKIPPED)
+        {
+            erase(currIt);
+        }
+        else
+        {
+            popTopTx(currIt);
+        }
+    }
+}
+
+std::pair<bool, int64_t>
+SurgePricingPriorityQueue::canFitWithEviction(
+    TransactionFrameBase const& tx, uint32_t txOpsDiscount,
+    std::vector<TxStackPtr>& txStacksToEvict) const
 {
-    return feeRate3WayCompare(l.getFeeBid(), l.getNumOperations(),
-                              r.getFeeBid(), r.getNumOperations());
+    // This only makes sense when the lowest fee rate tx is on top.
+    releaseAssert(!mComparator.isGreater());
+
+    if (tx.getNumOperations() < txOpsDiscount)
+    {
+        throw std::invalid_argument(
+            "Discount shouldn't be larger than transaction operations count");
+    }
+    auto txNewOps = tx.getNumOperations() - txOpsDiscount;
+    if (txNewOps > mOpsLimit)
+    {
+        CLOG_WARNING(
+            Herder,
+            "Operation limit is lower size than transaction ops: {} < {}."
+            "Node won't be able to accept all transactions.",
+            mOpsLimit, txNewOps);
+        return std::make_pair(false, 0ll);
+    }
+    uint32_t totalOps = sizeOps();
+    uint32_t newTotalOps = totalOps + txNewOps;
+    bool fitWithoutEviction = newTotalOps <= mOpsLimit;
+    // If there is enough space, return.
+    if (fitWithoutEviction)
+    {
+        return std::make_pair(true, 0ll);
+    }
+    auto iter = getTop();
+    int64_t neededTotalOps =
+        std::max<int64_t>(0, static_cast<int64_t>(newTotalOps) - mOpsLimit);
+    // The above checks should ensure there are some operations that need to be
+    // evicted.
+    releaseAssert(neededTotalOps > 0);
+    while (neededTotalOps > 0)
+    {
+        // The preconditions are ensured above that we should be able to fit
+        // the transaction by evicting some transactions (given high enough
+        // fee).
+        releaseAssert(iter != mTxStackSet.end());
+        auto const& evictTxStack = *iter;
+        auto const& evictTx = *evictTxStack->getTopTx();
+        auto evictOps = evictTxStack->getNumOperations();
```

Comment:
I think this method (`canFitWithEviction`) should not be in this class, or should be written differently:
what that code does is use the top tx `evictTx` to decide on evicting the entire stack (and the code in , which does not make sense in the general case, and only in the context of `TxQueueLimiter` that right now uses `SingleTxStack`.
If we want to keep it in here, the challenge (I think) is that you can't really iterate over txs in stacks without popping.

Reply:
> what that code does is use the top tx evictTx to decide on evicting the entire stack 
Well, that's the invariant of this method and it happens to work for `TxQueueLimiter`. I can't say it's incorrect for seq num based stack implementations, as the same invariant would be maintained (whether it's useful or not is up to the caller; currently I guess we know we don't need to reuse it).

I cannot say I'm 100% happy with this class, however I think it's a decent option. The set of supported operations is indeed bespoke and isn't too abstract. This is due to the fact that we have quite specific use cases and there is only so much in common between them. In order to support the upcoming functionality I wanted to encapsulate as much logic as possible in these methods and let as little internals to be exposed as possible. The implementation with lanes is more involved in terms of using the internal structures. 

I guess we could make the queue to be 'low-level' data structure that exposes the internals and then a wrapper data structure to be used in `TxQueueLimiter` with `SingleTxStack` only. However, I'm not sure how much value does that provide, given that the complexity still lies in the iteration methods. I'd also still rather keep the logic in `SurgePricingUtils` so that there is a single source for all surge-pricing related stuff.

Reply:
We talked offline about potentially getting rid of the "transaction chains" per account in the future, so I think it's fine to keep the code as is. As right now we have no use cases for this method outside of the `SingleTxStack`, and to avoid surprises if we were to use it in a different context, I think we should just add an assert that there is only one tx in the stack:
```
        auto const& evictTx = *evictTxStack->getTopTx();
        auto evictOps = evictTxStack->getNumOperations();
        releaseAssert(evictOps == evictTx.getNumOperations());
```

what do you think?

Reply:
SG, added the assertion.